File: jsyd_102/src/SSThreshhold.java
import java.io.*;

public class SSThreshhold extends SSModule implements SydConstants
{
  public ExpRec    cutOffExp = InitExp("0.5");


  public SSThreshhold(ModList mList)
  {
    super(MT_Threshhold, mList);
    DescribeLink(0, "Signal for Threshhold", "sig");
  }

  void  Copy(SSModule ss)
  {
    SSThreshhold sa = (SSThreshhold) ss;
    super.Copy(ss);
    CopyExp(sa.cutOffExp, cutOffExp);
  }

  void Save(PrintWriter ar) throws IOException
  {
    super.Save(ar);
    ar.println("THR " + cutOffExp.exp);
  }


  void Load(BufferedReader ar) throws IOException
  {
   cutOffExp = LoadExp(ar,"THR");
  }

	double GenerateOutput(SSModule callingMod)
	{
    double cutOff;
    cutOff = SolveExp(cutOffExp,callingMod);
    double retVal = MixInputs(-1,callingMod) >= cutOff? 1.0 : 0.0;
    lastRightSample = lastRightInput >= cutOff? 1.0 : 0.0;
    return retVal;
	}

  void Reset(SSModule callingMod)
  {
    super.Reset(callingMod);
    ResetExp(cutOffExp, callingMod);
  }

}

File: csyd_102/src/ssthreshhold.cpp
// ssthreshhold.cp

#include "ss.h"
#include "ssthreshhold.h"

SSThreshhold::SSThreshhold(ModList * mList, short h, short v) : SSModule(MT_Threshhold, mList, h, v)
{
	InitExp(&cutOffExp,"0.5");
	DescribeLink(0, "Signal for Threshhold", "sig",0,0,0xFFFF);
}

void	SSThreshhold::Copy(SSModule *ss)
{
	SSThreshhold *sa = (SSThreshhold *) ss;
	SSModule::Copy(ss);
	CopyExp(&sa->cutOffExp,&cutOffExp);
}

/*
void SSThreshhold::Save(FILE* ar)
{
	SSModule::Save(ar);
	WriteFileLine(ar,"THR %s\r",cutOffExp.exp);
}
*/

void SSThreshhold::Load(FILE* ar) 
{
	LoadExp(ar,"THR",&cutOffExp);
}


ssfloat SSThreshhold::GenerateOutput(SSModule *callingMod)
{
	ssfloat	retVal,cutOff;
	cutOff = SolveExp(&cutOffExp,callingMod);
	retVal = MixInputs(-1,callingMod) >= cutOff? 1 : 0;
  lastRightSample = lastRightInput >= cutOff? 1.0 : 0.0;
	return retVal;
}

void SSThreshhold::Reset(SSModule *callingMod)
{
	ResetExp(&cutOffExp,callingMod);
}

